drivers/flash/flash_mcux_flexspi_nor.c

```diff
@@ -0,0 +1,515 @@
+/*
+ * Copyright 2020 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT	nxp_imx_flexspi_nor
+
+#include <drivers/flash.h>
+#include <logging/log.h>
+#include <sys/util.h>
+#include "spi_nor.h"
+#include "flash_mcux_flexspi.h"
+
+#ifdef CONFIG_HAS_MCUX_CACHE
+#include <fsl_cache.h>
+#endif
+
+#define NOR_WRITE_SIZE	1
+#define NOR_ERASE_VALUE	0xff
+
+LOG_MODULE_DECLARE(flash_flexspi, CONFIG_FLASH_LOG_LEVEL);
+
+enum {
+	/* SPI instructions */
+	READ_ID,
+	READ_STATUS_REG,
+	WRITE_STATUS_REG,
+	WRITE_ENABLE,
+	PAGE_PROGRAM,
+	ERASE_SECTOR,
+	ERASE_CHIP,
+
+	/* Quad SPI instructions */
+	READ_FAST_QUAD_OUTPUT,
+	PAGE_PROGRAM_QUAD_INPUT,
+	ENTER_QPI,
+};
```

Comment:
Is the order of these commands important? Does it have to correspond to the look-up table index in Table 8-15 in the iMX RT1020 Reference Manual for example, for XIP to work in the future. Maybe I misunderstood how this works. It is a very flexible system, for good and bad.

Reply:
Based on the footnote after the table, I think we can change them:
```
1. All the pre-defined LUT indexes are only applicable to
boot stage. User application can use the whole 16 LUT
entries freely based on their requirement.
```

I ordered them this way because I think we may eventually want to conditionalize quad and octal support. Something like this:
```c
enum {
        /* SPI instructions */
        READ_ID,
        READ_STATUS_REG,
        WRITE_STATUS_REG,
        WRITE_ENABLE,
        PAGE_PROGRAM,
        ERASE_SECTOR,
        ERASE_CHIP,

#ifdef CONFIG_FLASH_FLEXSPI_NOR_QUAD
        /* Quad SPI instructions */
        READ_FAST_QUAD_OUTPUT,
        PAGE_PROGRAM_QUAD_INPUT,
        ENTER_QPI,
#endif

#ifdef CONFIG_FLASH_FLEXSPI_NOR_OCTAL
        /* Octal SPI instructions */
        READ_FAST_OCTAL_OUTPUT,
        PAGE_PROGRAM_OCTAL_INPUT,
        ENTER_OPI,
#endif
};

```